README.md
# qrize

[![Build Status](https://travis-ci.org/qrize/qrize.svg?branch=master)](https://travis-ci.org/qrize/qrize)
[![Coverage Status](https://coveralls.io/repos/github/qrize/qrize/badge.svg?branch=master)](https://coveralls.io/github/qrize/qrize?branch=master)

JavaScript library for generating **tiny** QR codes for your web pages.


## Table of Contents

- [Live demo](#live-demo)
- [Why qrize?](#why-qrize)
- [How to install?](#how-to-install)
- [How to use?](#how-to-use)
- [API](#api)
  - [Constructor](#constructor)
  - [Methods](#methods)
  - [Static Methods](#static-methods)
- [License](#license)


## Live demo

https://qrize.me


## Why qrize?

`qrize` integrates with URL shortener service, giving you tiny QR codes no matter how long a URL is.

Compare these QR codes that lead to the same [long address](https://www.amazon.com/b/ref=s9_acss_bw_en_BGG15eve_d_1_9_w?_encoding=UTF8&node=25&pf_rd_m=ATVPDKIKX0DER&pf_rd_s=merchandised-search-top-3&pf_rd_r=5NYMKAJQQQB3H3J1XTES&pf_rd_r=5NYMKAJQQQB3H3J1XTES&pf_rd_t=101&pf_rd_p=c0b03f4d-c947-45e3-9b66-2bd220509181&pf_rd_p=c0b03f4d-c947-45e3-9b66-2bd220509181&pf_rd_i=283155) (313 symbols):

| Regular QR code :hankey:                                                            | QR code you get with qrize :+1:                                                                |
|-------------------------------------------------------------------------------------|------------------------------------------------------------------------------------------------|
| ![Regular QR code](https://raw.githubusercontent.com/qrize/qrize/assets/big.png) | ![QR code you get with qrize](https://raw.githubusercontent.com/qrize/qrize/assets/small.png) |

Having QR codes smaller makes it more reader-friendly, saves space on page and is awesome. Visit [qrize.me](https://qrize.me) for general information.


## How to install?

To install qrize with Yarn, run:

```sh
yarn add qrize
```

To install qrize with npm, run:

```sh
npm install --save qrize
```

If you don’t want to use npm to manage client packages, `qrize` package also provides a single-file distribution, which is hosted on a CDN:

```html
<script src="https://unpkg.com/qrize/dist/qrize.umd.js"></script>
```

## How to use?

```js
import Qrize from "qrize";

const qrize = new Qrize({
  element: document.getElementById("qr-target")
});
qrize.createImg({ url: "http://example.com" });
```

That's it. An element with id `#qr-target` will contain `<img>` tag representing a QR code that leads to http://example.com.

If you use UMD package from CDN, `Qrize` constuctor is available globally.

## API

### Constructor

`Qrize` constuctor takes an options object as a parameter and these are available properties:

| Name     | Type        | Required | Description                                                                                                                                            |
|----------|-------------|----------|--------------------------------------------------------------------------------------------------------------------------------------------------------|
| element  | DOM Element | yes      | Container element for the generated QR code.                                                                                                           |
| cellSize | Number      | no       | Size of each grid cell in pixels. Default: `2`.                                                                                                        |
| margin   | Number      | no       | Size of margin around the QR code in pixels. Note that you might want to leave some space around your QR code in order to make it readable. Default: `0`. |

### Methods

Instantiated object returned by `new Qrize()` has three methods with pretty straightforward purposes:

- `createImg()`
- `createSvg()`
- `createTable()`

Each method can take (not requred) an options object with following parameters:

| Name      | Type     | Required | Description                                                                                                                                  |
|-----------|----------|----------|----------------------------------------------------------------------------------------------------------------------------------------------|
| url       | String   | no       | A URL that QR code will lead to. Defaults to a current page address. Default: `window.location.href`.                                        |
| onSuccess | Function | no       | A callback that will be called after successful QR code build. Is invoked with an object that has two properties: `hash` and provided `url`. |
| onFailure | Function | no       | A callback that will be called after an attempt to build a QR code has been failed.                                                          |

### Static Methods

There are also static `Qrize` methods that communicate with `qrize.me` API directly:

- `getHash()` - get unique hash id of the given URL. Options properties:

| Name      | Type     | Required | Description                                                                                                                              |
|-----------|----------|----------|------------------------------------------------------------------------------------------------------------------------------------------|
| url       | String   | yes      | A URL you want to get a hash id of.                                                                                                      |
| onSuccess | Function | no       | A callback that will be called on successful API response. Is invoked with an object that has two properties: `hash` and provided `url`. |
| onFailure | Function | no       | A callback that will be called if request failed. Is invoked with two arguments: error status code and error message.                    |

- `getUrl()` - get URL of the given hash id. Options properties:

| Name      | Type     | Required | Description                                                                                                                              |
|-----------|----------|----------|------------------------------------------------------------------------------------------------------------------------------------------|
| hash      | String   | yes      | A hash id you want to get a URL of.                                                                                                      |
| onSuccess | Function | no       | A callback that will be called on successful API response. Is invoked with an object that has two properties: `url` and provided `hash`. |
| onFailure | Function | no       | A callback that will be called if request failed. Is invoked with two arguments: error status code and error message.                    |

You probably won't need them, but if you will, you can use them like this:

```js
import Qrize from "qrize";

Qrize.getHash({
  url: 'http://example.com',
  onSuccess({ hash }) {
    console.log(hash);
  }
})

Qrize.getUrl({
  hash: '8jLDWGQ',
  onSuccess({ url }) {
    console.log(url);
  }
})
```

## LICENSE

MIT

QR Code is registered trademark of [DENSO WAVE INCORPORATED](http://www.denso-wave.com/en/).

http://www.qrcode.com/en/faq.html#patentH2Title


CHANGELOG.md
## 1.0.0 (February 21, 2023)

BREAKING CHANGE:
- Change API endpoint after migration to the DigitalOcean serverless functions


## 0.4.0 (January 8, 2017)

- Use https instead of http


## 0.3.1 (December 30, 2017)

- Add Live Demo link to the Readme


## 0.3.0 (December 27, 2017)

- onSuccess callback of `createImg`, `createSvg` and `createTable` methods receives an object with `hash` and `url` properties


## 0.2.1 (December 16, 2017)

- Add MIT licence


## 0.2.0 (December 16, 2017)

- Validate URL before sending to API

Internal:

- Tweak Prettier options


## 0.1.3 (November 26, 2017)

- Make instance of Qrize reusable


## 0.1.2 (November 13, 2017)

- Add keywords to package.json


## 0.1.1 (November 13, 2017)

- Fix typo in Readme


## 0.1.0 (November 11, 2017)

- Initial public release


flow-typed/qrcode-generator.js
declare module "qrcode-generator" {
  declare export type TypeNumber =
    | 0 // Automatic type number
    | 1
    | 2
    | 3
    | 4
    | 5
    | 6
    | 7
    | 8
    | 9
    | 10
    | 11
    | 12
    | 13
    | 14
    | 15
    | 16
    | 17
    | 18
    | 19
    | 20
    | 21
    | 22
    | 23
    | 24
    | 25
    | 26
    | 27
    | 28
    | 29
    | 30
    | 31
    | 32
    | 33
    | 34
    | 35
    | 36
    | 37
    | 38
    | 39
    | 40;

  declare export type ErrorCorrectionLevel = "L" | "M" | "Q" | "H";

  declare export type Mode =
    | "Numeric"
    | "Alphanumeric"
    | "Byte" /* Default */
    | "Kanji";

  declare export type QrCode = {|
    addData(data: string, mode?: Mode): void,
    make(): void,
    getModuleCount(): number,
    isDark(row: number, col: number): boolean,
    createImgTag(cellSize?: number, margin?: number): string,
    createSvgTag(cellSize?: number, margin?: number): string,
    createTableTag(cellSize?: number, margin?: number): string,
    renderTo2dContext(
      context: CanvasRenderingContext2D,
      cellSize?: number
    ): void,
  |};

  declare export default function qrcode(
    typeNumber: TypeNumber,
    errorCorrectionLevel: ErrorCorrectionLevel
  ): QrCode;
}
